### PR TITLE
[node] Allow augmentation of module

### LINKED ISSUE #19623: @types/node@8.0.27 breaks @types/uuid@3.4.2
- [/] I tried using the `@types/uuid` package and had problems.
- [-] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
 currently using typscript@2.3.4

our build fails since @types/uuid@3.4.2 loads the dependency @types/node@8.0.27, de build succeeded with @types/node@8.0.26 !

The Error we get is
`ERROR in [at-loader] ./node_modules/@types/node/index.d.ts:5615:11 
    TS2420: Class 'Module' incorrectly implements interface 'NodeModule'.
    Property 'hot' is missing in type 'Module'.`

### PR DESCRIPTION
**Pull request by Flarna**
- move class Module into NodeJS namespace
- add Module.Module
- remove `implements NodeModule` (fixes #19623)

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/nodejs/node/blob/447715543b5f80cbee6496657179f1393a631fbf/lib/module.js#L651
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by Flarna** [comment]
CI errors seem to be not related to my changes.

**Comment by devoto13** [comment]
It can be augment even without this change using [module augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) or am I missing something?

Also IMO we should aim at reducing amount of global types, not increasing them...

**Comment by Flarna** [comment]
@devoto13 I tried it but failed, but maybe the way I tried to do this is wrong. In general I agree that global types (even in namespace NodeJS) should be reduced if possible.

My local internal.ts holds following:
```
declare module "module" {
	namespace Module {
		function _resolveFilename(file: string, parent: NodeModule | undefined): string;
	}
}
```
I get error message: `error TS2309: An export assignment cannot be used in a module with other exported elements.`
To my understanding the diff to the sample on typescript doc is that here `export =` is used.

If I add `import "module"` on top of internal.ts I get `error TS2671: Cannot augment module 'module' because it resolves to a non-module entity.`

With my proposed change above I can do this by adding the `namespace Module` inside `namespace NodeJS` in my local file.

I use typescript 2.5.2. Any hints welcome.

**Comment by kjin** [comment]
Thanks for doing this! I'm in need of the same functionality.

Would it be possible to reflect example usage in the test file? Like appending

```ts
declare global {
  namespace NodeJS {
    namespace Module {
      function _resolveFilename(file: string, parent: NodeModule | undefined): string;
    }
  }
}
```

to the `module` module tests and then using `_resolveFilename` immediately afterwards.

**Comment by devoto13** [comment]
@Flarna I spent some time today researching this issue. And the main problem is indeed usage of `export =`. 

The issue is that in this case module namespace (which is available to augmenting declaration) becomes the exported symbol itself making symbols declared inside original module not visible inside augmenting declaration. In this case `Module` class becomes a module namespace. From TypeScript point of view `class` is not a `module`, hence `error TS2671: Cannot augment module 'module' because it resolves to a non-module entity.`. This can be [workarounded](https://github.com/Microsoft/TypeScript/issues/6656#issuecomment-175780097) by adding an empty `namespace/module` with the same name in the original module:

```
declara module "module" {
  class Module { ... }
  namespace Module { }
  export = Module;
}
```

In augmenting declaration `Module` symbol is not visible, because it gets `Module` class/namespace declared above as its root namespace. It allows to add static member using function/variables declarations (unexpected syntax), but I didn't manage to find a way to add instance members.

```
declare module "module" {
  // Won't work since this namespace is inside `Module` class.
  interface Module {
    newFancyField: number;
  }

  let myVar: number; // adds static `myVar` to `Module` class
  function test(): void; // adds static `test` function to `Module` class
}
```

FWIW moving class to the global namespace is probably the best option at the moment.

File: types/node/index.d.ts [hunk 1]
```diff
@@ -671,6 +671,24 @@ declare namespace NodeJS {
...
+        static wrap(code: string): string;
+
+        static Module: typeof Module;
+
+        exports: any;
+        require: NodeRequireFunction;
+        id: string;
+        filename: string;
+        loaded: boolean;
+        parent: NodeModule | null;
```
Comment (devoto13) [inline types/node/index.d.ts]: Can you also change this and below from `NodeModule` to `Module`?

I believe the end goal will be to replace all usages of `NodeModule` with `Module`.

**Review by devoto13 — APPROVED** [review]

File: types/node/index.d.ts (hunk 1 above)
Comment (Flarna) [reply types/node/index.d.ts]: done. I remove also the `implements NodeModule` as it caused issues with other modules which augment `NodeModule` and as you indicate the target is anyhow to get rid of it.

**Review by devoto13 — APPROVED** [review]

**Comment by devoto13** [comment]
@andy-ms can we get this PR in? Travis failures don't seem to be related to these changes. It will allow people to augment both `NodeModule` and `Module` without running into issues like https://github.com/DefinitelyTyped/DefinitelyTyped/issues/19623.

**Comment by andy-ms** [comment]
CC @RyanCavanaugh -- is there a way to augment an  `export =` module?

**Comment by devoto13** [comment]
To be precise the problem only exists when trying to augment `class` or `interface` exported with `export =`. It's not a problem to augment anything inside a namespace exported using `export =`. 

See [this test case](https://github.com/Microsoft/TypeScript/blob/master/tests/cases/compiler/augmentExportEquals6_1.ts) for an example. Class `A` can be easily augment, but there doesn't seem to be a way to augment class `foo`.

**Comment by Flarna** [comment]
Merge master again but CI still fails because of unrelated issues (e.a. #19740).
@andy-ms Any chance that PRs for node get merged even CI fails like for this issue?

**Comment by Flarna** [comment]
@andy-ms, @mhegazy Sorry to ping but is there anything I can do to move this PR forward?

**Pull request merged**